stdlib/public/TensorFlow/ShapedArray.swift

```diff
@@ -264,14 +264,174 @@ internal extension _ShapedArrayProtocol {
   }
 }
 
-/// Common public protocol implementations
+fileprivate extension String {
+  /// Returns a string of the specified length, padded with whitespace to the
+  /// left.
+  func leftPadded(toLength length: Int) -> String {
+    return repeatElement(" ", count: Swift.max(0, length - count)) + self
+  }
+}
+
+/// Common public protocol implementations.
+
 fileprivate extension _ShapedArrayProtocol
-  where Element : _ShapedArrayProtocol {
-  var _description: String {
+  where Element : _ShapedArrayProtocol, Element == Element.Element
+{
+  /// Returns the whitespace separator between elements, given the current
+  /// indent level.
+  func separator(indentLevel: Int) -> String {
+    if rank == 1 {
+      return ", "
+    }
+    return String(repeating: "\n", count: rank - 1) +
+           String(repeating: " ", count: indentLevel + 1)
+  }
+
+  /// A textual representation of the 1-D shaped array, starting at the given
+  /// indent level. Returns a summarized description if `summarize` is true and
+  /// the element count exceeds twice the `edgeElementCount`.
+  ///
+  /// - Parameters:
+  ///   - indentLevel: The indentation level.
+  ///   - edgeElementCount: The maximum of elements to print before and after
+  ///     summarization via ellipses (`...`).
+  ///   - maxScalarLength: The length of the longest scalar description in the
+  ///     entire original array-to-print.
+  ///   - scalarCountPerLine: The number of scalars to print per line, used when
+  ///     printing 1-D vectors.
+  ///   - summarize: If true, summarize description if element count exceeds
+  ///     twice `edgeElementCount`.
+  func vectorDescription(
+    indentLevel: Int, edgeElementCount: Int, maxScalarLength: Int,
+    scalarCountPerLine: Int, summarize: Bool
+  ) -> String {
+    // Get scalar descriptions.
+    func scalarDescription(_ element: Element) -> String {
+      let description = String(describing: element)
+      return description.leftPadded(toLength: maxScalarLength)
+    }
+    var scalarDescriptions: [String] = []
+    if summarize && count > 2 * edgeElementCount {
+      scalarDescriptions +=
+        prefix(edgeElementCount).map(scalarDescription)
+      scalarDescriptions += ["..."]
+      scalarDescriptions +=
+        suffix(edgeElementCount).map(scalarDescription)
+    } else {
+      scalarDescriptions += map(scalarDescription)
+    }
+
+    // Combine scalar descriptions into lines, based on the scalar count per
+    // line.
+    let lines = stride(
+      from: 0, to: scalarDescriptions.count, by: scalarCountPerLine
```

Comment:
It would be more principled to use `scalarDescriptions .startIndex` and `scalarDescriptions .endIndex` instead of `0` and `scalarDescriptions .count`.